iothub_client/samples/pnp/pnp_simple_thermostat/pnp_simple_thermostat.c

```diff
@@ -0,0 +1,441 @@
+    {
+        LogError("Unable to send reported state for target temperature.  Error=%d", iothubClientResult);
+    }
+    else
+    {
+        LogInfo("Sending maxTempSinceReboot property");
+    }
+}
+
+//
+// SendMaxTemperatureSinceReboot reports a PnP property indicating the maximum temperature since the last reboot (simulated here by lifetime of executable)
+//
+static void SendMaxTemperatureSinceReboot(IOTHUB_DEVICE_CLIENT_HANDLE deviceClient)
+{
+    IOTHUB_CLIENT_RESULT iothubClientResult;
+    char maxTemperatureSinceRebootProperty[256];
+
+    if (snprintf(maxTemperatureSinceRebootProperty, sizeof(maxTemperatureSinceRebootProperty), g_maxTemperatureSinceRebootFormat, g_maxTemperature) < 0)
+    {
+        LogError("snprintf building maxTemperature failed");
+    }
+    else if ((iothubClientResult = IoTHubDeviceClient_SendReportedState(deviceClient, (const unsigned char*)maxTemperatureSinceRebootProperty, strlen(maxTemperatureSinceRebootProperty), NULL, NULL)) != IOTHUB_CLIENT_OK)
+    {
+        LogError("Unable to send reported state for maximum temperature.  Error=%d", iothubClientResult);
+    }
+    else
+    {
+        LogInfo("Sending maxTempSinceReboot property");
+    }
+}
+
+//
+// Thermostat_DeviceTwinCallback is invoked by the IoT SDK when a twin - either full twin or a PATCH update - arrives.
+//
+static void Thermostat_DeviceTwinCallback(DEVICE_TWIN_UPDATE_STATE updateState, const unsigned char* payload, size_t size, void* userContextCallback)
+{
+    // The device handle associated with this request is passed as the context, since we will need to send reported events back.
+    IOTHUB_DEVICE_CLIENT_HANDLE deviceClient = (IOTHUB_DEVICE_CLIENT_HANDLE)userContextCallback;
+
+    char* jsonStr = NULL;
+    JSON_Value* rootValue = NULL;
+    JSON_Object* desiredObject;
+    JSON_Value* versionValue = NULL;
+    JSON_Value* targetTemperatureValue = NULL;
+
+    if ((jsonStr = CopyTwinPayloadToString(payload, size)) == NULL)
+    {
+        LogError("Unable to allocate twin buffer");
+    }
+    else if ((rootValue = json_parse_string(jsonStr)) == NULL)
+    {
+        LogError("Unable to parse twin JSON");
+    }
+    else if ((desiredObject = GetDesiredJson(updateState, rootValue)) == NULL)
+    {
+        LogError("Cannot retrieve desired JSON object");
+    }
+    else if ((targetTemperatureValue = json_object_get_value(desiredObject, PnP_JsonTargetTemperature)) == NULL)
+    {
+        // The target temperature not being set is NOT an error, but it does indicate we can't proceed any further.
+        ;
+    }
+    else if ((versionValue = json_object_get_value(desiredObject, PnP_JsonPropertyVersion)) == NULL)
+    {
+        // The $version does need to be set in *any* legitimate twin desired document.  Its absence suggests 
+        // something is fundamentally wrong with how we've received the twin and we should not proceed.
+        LogError("Cannot retrieve field %s for twin", PnP_JsonPropertyVersion);
```

Comment:
nit:
```c
LogError("Cannot retrieve field %s for twin", PnP_JsonPropertyVersion);
```
I'm imagining I am a customer using this and I get this error. I'd be confused why `$version` couldn't be found even though I didn't send it. Maybe a little addition like
```suggestion
        LogError("Cannot retrieve field %s for underlying twin protocol", PnP_JsonPropertyVersion);
```
to make it a little clear it's purpose?

Reply:
Yeah, good feedback.  I've rewritten even a bit more verbosily for you to check out on next round.